Upgrade travis platforms

vadz opened the pull request:
As discussed in [this thread](https://groups.google.com/forum/#!topic/wx-dev/t6Cwco0D6Ho)

vadz committed:
Don't rebuild on AppVeyor if only .travis.yml was changed

Just avoid the unnecessary rebuilds.

vadz committed:
Add Travis CI build using Xcode 11.3

Test with the latest Xcode release too.

Make this build monolithic just to try to cover more possibilities.

vadz committed:
Upgrade Travis CI Linux build platforms to later versions

Move Precise (12.04) build to Trusty (4.04), move one of the existing
Trusty builds to Xenial (16.04) and all the other ones to Bionic (18.04)
to test under reasonably recent platforms too, and mostly.

vadz committed:
Disable debug info generation in Travis CI builds

Debug info is useless in these builds and disabling it should hopefully
speed them up a bit and take less space.

vadz committed:
Show config.log contents if configure failed

This should help understanding why Qt libraries are not found under
Bionic and may also be useful in case of other failures in the future.

vadz committed:
Ignore more Travis CI files on AppVeyor

Avoid unnecessary AppVeyor rebuilds when Travis build files are changed.

vadz committed:
Use C++17 and STL for Xcode 11.3 Travis CI build

Note that using C++98 results in tests build failure with clang as
libc++ doesn't allow using custom std::swap() specialization for
non-const references in its iter_swap(), it really requires using rvalue
references, which are, of course, only available in C++11 and later.

paulcor commented on build/tools/travis-ci.sh:
```diff
@@ -54,7 +54,14 @@ case $wxTOOLSET in
         echo 'Configuring...' && echo -en 'travis_fold:start:script.configure\\r'
-        ./configure --disable-optimise $wxCONFIGURE_FLAGS
+        ./configure --disable-optimise --disable-debug_info $wxCONFIGURE_FLAGS || rc=$?
+        if [ $rc != 0 ]; then
```
`build/tools/travis-ci.sh: line 58: [: !=: unary operator expected`

vadz replied:
Sorry, could you please explain where do you see this? I must be blind, but I don't see anything wrong with this test...

paulcor replied:
line 1305: [https://travis-ci.org/github/wxWidgets/wxWidgets/jobs/678182859](url)

`rc` is not set if configure return code is 0.

vadz replied:
Thanks, I've committed an ugly fix in 3fc119f564 (there must be a better way to do it, but somehow I don't see it at the moment...).

vadz committed:
Install Qt OpenGL support when building wxQt on Travis

Somehow this wasn't necessary on the old Ubuntu version, but is needed
under Bionic.

Ideally configure shouldn't refuse to build everything if just Qt5OpenGL
is not found, and just disable OpenGL support, but this enhancement can
be done later.

vadz committed:
Disable AppVeyor rebuilds on Mac files changes too

As long as we don't test under AppVeyor, there is no need to do anything
there when only Mac-specific files are changed.

vadz committed:
Require macOS 10.12 for Travis CI build using Xcode 11.3

Our tests don't compile under Mac when targeting 10.9 and using C++17 as
they try to use C++17 std::uncaught_exceptions() which is only available
since 10.12 under macOS.

We could avoid the use of this function in the tests in this build
configuration, but for now just target 10.12 instead. Yet another
alternative would be to downgrade to using C++14.

vadz committed:
Don't install unnecessary packages in Travis CI build script

It shouldn't be necessary to install GTK-specific packages when building
non-GTK-based ports.

vadz committed:
Avoid using undefined variable in Travis CI build script

rc is undefined if configure succeeds, so use "$rc" to expand to
something non-empty even in this case.

MaartenBent commented on .travis.yml:
```diff
@@ -41,22 +41,26 @@ matrix:
           compiler: clang
           env: wxTOOLSET=cmake wxCMAKE_GENERATOR=Xcode wxCMAKE_DEFINES="-DCMAKE_CXX_STANDARD=11"
           name: wxOSX Xcode 9.4
```
```suggestion
          name: wxOSX CMake Xcode 9.4
```

vadz replied:
Thanks, done.

paulcor commented on build/tools/travis-ci.sh:
```diff
@@ -54,7 +54,14 @@ case $wxTOOLSET in
         echo 'Configuring...' && echo -en 'travis_fold:start:script.configure\\r'
-        ./configure --disable-optimise $wxCONFIGURE_FLAGS
+        ./configure --disable-optimise --disable-debug_info $wxCONFIGURE_FLAGS || rc=$?
+        if [ "$rc" != 0 ]; then
```
Still not right. How about this:
```
./configure --disable-optimise --disable-debug_info $wxCONFIGURE_FLAGS
rc=$?
if [ $rc != 0 ]; then
```
```

paulcor replied:
Ahh, no, there is a `set -e`. Maybe this:
```
if [ -n "$rc" ]; then
```

vadz replied:
Yes, this is what I ended up by doing, sorry for the flurry of commits.

I still have no idea why does GTK 3 STL build fail though.

vadz committed:
Really fix configure exit code test

Don't compare undefined variable with 0, this is not going to work.

vadz committed:
Change wxOSX CMake build name to include "CMake"

vadz committed:
Install libpango1.0-dev for wxX11 build on Travis

Hopefully this is what explains pangoxft pkg-config check failure.

vadz committed:
Fix configure success test one more time

Quotes are important.

vadz committed:
Disable GUI parts build for PPC and S/390 Travis platforms

People are unlikely to use wxGTK on those anyhow, and it takes an
awfully long time to build, especially on PPC.

MaartenBent commented:
I can reproduce the GTK3 STL problem locally. 

Should `libwx_baseu_xml` be in `objdump -p test_gui`?
```
Dynamic Section:
  NEEDED               libwx_gtk3u_stc-3.1.so.4
  NEEDED               libwx_gtk3u_aui-3.1.so.4
  NEEDED               libwx_gtk3u_richtext-3.1.so.4
  NEEDED               libglib-2.0.so.0
  NEEDED               libwx_gtk3u_xrc-3.1.so.4
  NEEDED               libwx_gtk3u_html-3.1.so.4
  NEEDED               libwx_gtk3u_core-3.1.so.4
  NEEDED               libwx_baseu_net-3.1.so.4
  NEEDED               libwx_baseu-3.1.so.4
  NEEDED               libm.so.6
  NEEDED               libstdc++.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libpthread.so.0
  NEEDED               libc.so.6
  RUNPATH              /home/maarten/build/wxWidgets/cfg/lib
...
```

`ldd test_gui`:
```
	linux-vdso.so.1 (0x00007ffd5c7be000)
	libwx_gtk3u_stc-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_gtk3u_stc-3.1.so.4 (0x00007f15da368000)
	libwx_gtk3u_aui-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_gtk3u_aui-3.1.so.4 (0x00007f15da078000)
	libwx_gtk3u_richtext-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_gtk3u_richtext-3.1.so.4 (0x00007f15d9c18000)
	libglib-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007f15d9901000)
	libwx_gtk3u_xrc-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_gtk3u_xrc-3.1.so.4 (0x00007f15d95a6000)
	libwx_gtk3u_html-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_gtk3u_html-3.1.so.4 (0x00007f15d9268000)
	libwx_gtk3u_core-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_gtk3u_core-3.1.so.4 (0x00007f15d8688000)
	libwx_baseu_net-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_baseu_net-3.1.so.4 (0x00007f15d8428000)
	libwx_baseu-3.1.so.4 => /home/maarten/build/wxWidgets/cfg/lib/libwx_baseu-3.1.so.4 (0x00007f15d7ee9000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f15d7b4b000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f15d77c2000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f15d75aa000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f15d738b000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f15d6f9a000)
	libgdk-3.so.0 => /usr/lib/x86_64-linux-gnu/libgdk-3.so.0 (0x00007f15d6ca4000)
	libgtk-3.so.0 => /usr/lib/x86_64-linux-gnu/libgtk-3.so.0 (0x00007f15d639c000)
	libgobject-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0 (0x00007f15d6148000)
	libwx_baseu_xml-3.1.so.4 => not found
	libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f15d5ed6000)
	libwx_baseu_xml-3.1.so.4 => not found
	libpangocairo-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangocairo-1.0.so.0 (0x00007f15d5cc9000)
...
```

vadz committed:
Update Travis testing script to work for non-GUI builds too

Skip parts that don't work in this case.

vadz commented:
> I can reproduce the GTK3 STL problem locally.

I can't, but I don't have Bionic VM yet...

> Should `libwx_baseu_xml` be in `objdump -p test_gui`?

It should be in `objdump -p` output, of course, the question is why isn't it? The linker command line includes it here. But then I do see it in `objdump` output too (FWIW I prefer the output from `readelf -Wd` but they show the same thing anyhow).

Anyhow, it looks like `libwx_baseu_xml` is not linked into the test, but as it's required by one of the libraries that it links with, it is then not found during run-time. But I just don't understand why. Could you please check if it's present in the link command line for you? TIA!

MaartenBent commented:
Yes, it's there. It happens also with just `configure` or `configure --with-gtk=2`. And also with 3.1.3.
```
g++ -std=gnu++11 -o test_gui  test_gui_asserthelper.o test_gui_test.o test_gui_testableframe.o test_gui_rect.o test_gui_size.o test_gui_point.o test_gui_region.o test_gui_bitmap.o test_gui_colour.o test_gui_ellipsization.o test_gui_measuring.o test_gui_affinematrix.o test_gui_boundingbox.o test_gui_clippingbox.o test_gui_graphmatrix.o test_gui_graphpath.o test_gui_config.o test_gui_auitest.o test_gui_bitmapcomboboxtest.o test_gui_bitmaptogglebuttontest.o test_gui_bookctrlbasetest.o test_gui_buttontest.o test_gui_checkboxtest.o test_gui_checklistboxtest.o test_gui_choicebooktest.o test_gui_choicetest.o test_gui_comboboxtest.o test_gui_dataviewctrltest.o test_gui_datepickerctrltest.o test_gui_frametest.o test_gui_gaugetest.o test_gui_gridtest.o test_gui_headerctrltest.o test_gui_htmllboxtest.o test_gui_hyperlinkctrltest.o test_gui_itemcontainertest.o test_gui_label.o test_gui_listbasetest.o test_gui_listbooktest.o test_gui_listboxtest.o test_gui_listctrltest.o test_gui_listviewtest.o test_gui_markuptest.o test_gui_notebooktest.o test_gui_ownerdrawncomboboxtest.o test_gui_pickerbasetest.o test_gui_pickertest.o test_gui_radioboxtest.o test_gui_radiobuttontest.o test_gui_rearrangelisttest.o test_gui_richtextctrltest.o test_gui_searchctrltest.o test_gui_simplebooktest.o test_gui_slidertest.o test_gui_spinctrldbltest.o test_gui_spinctrltest.o test_gui_styledtextctrltest.o test_gui_textctrltest.o test_gui_textentrytest.o test_gui_togglebuttontest.o test_gui_toolbooktest.o test_gui_treebooktest.o test_gui_treectrltest.o test_gui_treelistctrltest.o test_gui_virtlistctrltest.o test_gui_webtest.o test_gui_windowtest.o test_gui_dialogtest.o test_gui_clone.o test_gui_evtlooptest.o test_gui_propagation.o test_gui_keyboard.o test_gui_exec.o test_gui_fonttest.o test_gui_image.o test_gui_rawbmp.o test_gui_htmlparser.o test_gui_htmlwindow.o test_gui_htmprint.o test_gui_accelentry.o test_gui_menu.o test_gui_guifuncs.o test_gui_selstoretest.o test_gui_garbage.o test_gui_safearrayconverttest.o test_gui_settings.o test_gui_socket.o test_gui_tlw.o test_gui_dataview.o test_gui_rowheightcachetest.o test_gui_boxsizer.o test_gui_gridsizer.o test_gui_wrapsizer.o test_gui_toplevel.o test_gui_valnum.o test_gui_valtext.o test_gui_clientsize.o test_gui_setsize.o test_gui_xrctest.o     -L/home/maarten/build/wxWidgets/cfgtr/lib -Wl,-rpath,/home/maarten/build/wxWidgets/cfgtr/lib    -pthread   -lwx_gtk3u_stc-3.1 -lwx_gtk3u_aui-3.1  -lwx_gtk3u_richtext-3.1  -lwx_gtk3u_media-3.1 -lgstvideo-1.0 -lgstbase-1.0 -lgstreamer-1.0 -lgobject-2.0 -lglib-2.0 -lwx_gtk3u_xrc-3.1  -lwx_baseu_xml-3.1 -lexpat -lwx_gtk3u_html-3.1  -lwx_gtk3u_core-3.1  -lwx_baseu_net-3.1  -lwx_baseu-3.1        -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lgthread-2.0 -pthread -lglib-2.0 -lX11 -lXxf86vm -lSM -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lnotify -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lXtst -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lpng -lz -ljpeg -ltiff -llzma -ljbig  -lwxregexu-3.1  -pthread   -Wl,--version-script,/home/maarten/build/wxWidgets/cfgtr/version-script -lz -ldl -lsecret-1 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -llzma -lm  -lz -ldl -lsecret-1 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -llzma -lm

```

vadz commented:
Thanks, but unfortunately this doesn't help and I still just don't see what could explain this... I'll have to install Bionic here to try to debug what's going on.

Note that it does work in GTK 2 build on Travis.

MaartenBent commented:
> Note that it does work in GTK 2 build on Travis.

Those are monolithic and static.

MaartenBent commented:
It also happens to xrcdemo
`
/xrcdemo: error while loading shared libraries: libwx_gtk3u_html-3.1.so.4: cannot open shared object file: No such file or directory
`

vadz commented:
OK, using `-Wl,--disable-new-dtags` provides a workaround. But what's the new way of doing this correctly?

vadz commented:
IIUC the problem is that ld 2.30, used under Bionic, doesn't consider executable rpath when searching for the indirect dependencies, i.e. dependencies of the shared libraries used by the executable. This makes sense and probably means that we need to build our libraries themselves with the correct rpath, so that they find their dependencies in the same directory.

Except it doesn't explain why this linker doesn't record the dependency in the executable. Especially when ld 2.31 does do it.

vadz commented:
And the answer to the last question is that Bionic gcc is built with `--as-needed` in its `link` spec for whatever reason. so using `-Wl,--no-as-needed` works too. I'll just do this for now, to be able to (hopefully) finally merge this PR, but a better solution will be to link our shared libraries with correct rpath, of course.

vadz committed:
Use ld --no-as-needed in Travis Ubuntu 18.04 wxGTK build

This works around a problem in our build system which doesn't set rpath
correctly when linking our own libraries, which means that "qa" library,
for example, doesn't record the path to its "xml" library dependency.

We already worked around this by linking any program using "qa" with
"xml" too, however this doesn't do anything if "--as-needed" flag is in
effect, as is the case by default under Ubuntu 18.04, so make our
workaround work by forcefully disabling this flag.

Real solution would be to use -Wl,-rpath correctly and then stop linking
with "xml" library unnecessarily in e.g. test_gui bakefile.
